### PR TITLE
[SPARK-8774] [ML] Add R model formula with basic support as a transformer

### PR DESCRIPTION
**Pull request by ericl**
This implements minimal R formula support as a feature transformer. Both numeric and string labels are supported, but features must be numeric for now.

cc @mengxr

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import org.apache.spark.ml.param.{Param, ParamMap}
+import org.apache.spark.ml.param.shared.{HasFeaturesCol, HasLabelCol}
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+/**
+ * :: Experimental ::
+ * Implements the transforms required for fitting a dataset against a R model formula.
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Also mention the operators we support.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ */
+@Experimental
+private[spark] class RModelFormula(override val uid: String)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Remove `private[spark]` so Scala users can also use it.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  extends Transformer with HasFeaturesCol with HasLabelCol {
+
+  def this() = this(Identifiable.randomUID("rModelFormula"))
+
+  val formula: Param[String] = new Param(this, "formula", "R model formula")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Missing doc and `@group param` in the ScalaDoc. The group is used to group methods in the generated Scala doc.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  protected var parsedFormula: Option[RFormula] = None
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Why is this `protected` instead of `private`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+  /**
+   * Sets the formula to use for this transformer. Must be called before use.
+   * @param value a R formula in string form (e.g. "y ~ x + z")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: * missing `@group setParam`
* `a R` -> `an R`
* missing `getFormula`

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+   */
+  def setFormula(value: String): this.type = {
+    parsedFormula = Some(RFormulaParser.parse(value))
+    set(formula, value)
+    this
+  }
+
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Missing setters for `featuresCol` and `labelCol`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  override def transform(dataset: DataFrame): DataFrame = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    transformLabel(featureTransformer.transform(dataset))
+  }
+
+  override def copy(extra: ParamMap): RModelFormula = defaultCopy(extra)
+
+  override def toString: String = s"RModelFormula(${get(formula)})"
+
+  protected def transformLabel(dataset: DataFrame): DataFrame = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: private?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    val responseName = parsedFormula.get.response
+    dataset.schema(responseName).dataType match {
+      case _: NumericType | BooleanType =>
+        dataset.select(
+          col("*"),
+          dataset(responseName).cast(DoubleType).as($(labelCol)))
+      case StringType =>
+        new StringIndexer(uid)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Should use a random uid.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  protected def featureTransformer: Transformer = {
+    // TODO(ekl) add support for non-numeric features and feature interactions
+    new VectorAssembler(uid)
+      .setInputCols(parsedFormula.get.terms.toArray)
+      .setOutputCol($(featuresCol))
+  }
+}
+
+/**
+ * :: Experimental ::
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: We don't need `:: Experimental ::` on private classes.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+ * Represents a parsed R formula.
+ */
+private[ml] case class RFormula(response: String, terms: Seq[String])
+
+/**
+ * :: Experimental ::
+ * Limited implementation of R formula parsing. Currently supports: '~', '+'.
+ */
+private[ml] object RFormulaParser extends RegexParsers {
+  def term: Parser[String] = "([a-zA-Z]|\\.[a-zA-Z_])[a-zA-Z0-9._]*".r
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Does R accept `$` in terms?

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala [hunk 11]
```diff
@@ -116,7 +116,7 @@ class VectorAssembler(override val uid: String)
     if (schema.fieldNames.contains(outputColName)) {
       throw new IllegalArgumentException(s"Output column $outputColName already exists.")
     }
-    StructType(schema.fields :+ new StructField(outputColName, new VectorUDT, false))
+    StructType(schema.fields :+ new StructField(outputColName, new VectorUDT, true))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala]: Is this change necessary? We always assume that the vector is always available.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.spark.mllib.util.TestingUtils._
+
+class RFormulaModelSuite extends SparkFunSuite with MLlibTestSparkContext {
+  test("params") {
+    ParamsSuite.checkParams(new RModelFormula())
+  }
+
+  test("parse simple formulas") {
+    def check(formula: String, response: String, terms: Seq[String]) {
+      new RModelFormula().setFormula(formula)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: Should it be in a separate test?

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    val resultSchema = formula.transformSchema(original.schema)
+    val expected = sqlContext.createDataFrame(
+      Seq(
+        (0, 1.0, 3.0, Vectors.dense(Array(1.0, 3.0)), 0.0),
+        (2, 2.0, 5.0, Vectors.dense(Array(2.0, 5.0)), 2.0))
+      ).toDF("id", "v1", "v2", "features", "label")
+    assert(result.schema.toString == resultSchema.toString)
+    assert(resultSchema.toString == expected.schema.toString)
+    assert(
+      result.collect.map(_.toString).mkString(",") ==
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: `collect` -> `collect()` (because it is an action). `collect` doesn't really guarantee the ordering. So it would be nice to put the expected result along with the input data as extra columns. Then make assertions on each record.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 2 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 1 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 3 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 4 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 5 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 6 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 7 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 8 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 9 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 10 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Looks like R supports arbitrary expressions in terms, so we'd need a full parser to be sure. For $ am I not sure it makes sense, since we assume the terms are from the dataframe already.

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala (hunk 11 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala]: I noticed that the schema of transform() has it as nullable, so probably transformSchema() should also. One alternative is to make transform() mark the vector as non-null, but I am not exactly sure how to do that.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 12 above)
Comment (ericl) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: I put it here since the parser is basically private to RModelFormula but could be convinced otherwise.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 13 above)
Comment (ericl) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: Do you know the right way to compare schemas / Rows for equality? It seems equals() is not implemented for either.

Also added sorted to fix the ordering issue.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+/**
+ * :: Experimental ::
+ * Implements the transforms required for fitting a dataset against an R model formula. Currently
+ * we support a limited subset of the R operators, including '~' and '+'. Also see the R formula
+ * docs here: http://www.inside-r.org/r-doc/stats/formula
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Use `http://stat.ethz.ch/R-manual/R-patched/library/stats/html/formula.html` instead, which is in the raw R manual format.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    parsedFormula = Some(RFormulaParser.parse(value))
+    set(formula, value)
+    this
+  }
+
+  /** @group getParam */
+  def getFormula: String = $(formula)
+
+  /** @group getParam */
+  def setFeaturesCol(col: String): this.type = set(featuresCol, col)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: `col` -> `value` (to be consistent with other setters. Since the method name already contains this info, it is not necessary to repeat that for the arg, especially for really long names)

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    val nullable = schema(parsedFormula.get.response).dataType match {
+      case _: NumericType | BooleanType => false
+      case _ => true
+    }
+    StructType(withFeatures.fields :+ StructField($(labelCol), DoubleType, nullable))
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    transformLabel(featureTransformer.transform(dataset))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: To be consistent, rename `featureTransformer` to `transformFeatures`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  override def copy(extra: ParamMap): RModelFormula = defaultCopy(extra)
+
+  override def toString: String = s"RModelFormula(${get(formula)})"
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: minor: `${get(formula))` -> `$getFormula` (slightly easier to read)

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+  private def transformLabel(dataset: DataFrame): DataFrame = {
+    val responseName = parsedFormula.get.response
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: "response", "target", or "label" are all valid names. In MLlib, we use "label". So it might be useful to rename "response" to "label".

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    dataset.schema(responseName).dataType match {
+      case _: NumericType | BooleanType =>
+        dataset.select(
+          col("*"),
+          dataset(responseName).cast(DoubleType).as($(labelCol)))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: What if the `responseName` is the same as `labelCol`? This may cause unexpected behavior. If the input is `DoubleType`, we should allow `labelCol` be the same as the target term in the formula. If we need to do transformation, then user should set a different `labelCol`. We can set the default `featuresCol` and `labelCol` based on the uid and hence it won't have name collision. I don't think this is a good solution, but I don't have good suggestions.

Btw, we can use `DataFrame.withColumn` to append a new column.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+class RFormulaModelSuite extends SparkFunSuite with MLlibTestSparkContext {
+  test("params") {
+    ParamsSuite.checkParams(new RModelFormula())
+  }
+
+  test("parse simple formulas") {
+    def check(formula: String, response: String, terms: Seq[String]) {
+      new RModelFormula().setFormula(formula)
+      val parsed = RFormulaParser.parse(formula)
+      assert(parsed.response == response)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: use `===` instead of `==` (and please update other `==`s)

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    val original = sqlContext.createDataFrame(
+      Seq((0, 1.0, 3.0), (2, 2.0, 5.0))).toDF("id", "v1", "v2")
+    val result = formula.transform(original)
+    val resultSchema = formula.transformSchema(original.schema)
+    val expected = sqlContext.createDataFrame(
+      Seq(
+        (0, 1.0, 3.0, Vectors.dense(Array(1.0, 3.0)), 0.0),
+        (2, 2.0, 5.0, Vectors.dense(Array(2.0, 5.0)), 2.0))
+      ).toDF("id", "v1", "v2", "features", "label")
+    assert(result.schema.toString == resultSchema.toString)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: Maybe it is worth leaving a TODO here for `DataType.equals`.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    assert(resultSchema.toString == expected.schema.toString)
+    assert(
+      result.collect().map(_.toString).sorted.mkString(",") ==
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: I don't think we need `toString` and `mkString(",")`. Maybe `sorted` is not necessary either.

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala (hunk 11 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala]: Okay, I think this is minor.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.ml.param.ParamsSuite
+import org.apache.spark.mllib.linalg.Vectors
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.mllib.util.TestingUtils._
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: remove unused imports

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+
+class RFormulaModelSuite extends SparkFunSuite with MLlibTestSparkContext {
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: `RFormulaModelSuite` -> `RModelFormulaSuite`

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 21 above)
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: Just figured out why. The column output from `VectorAssembler` also contains ML attributes that stores feature names. It is not included in `toString` ... If you compare the JSON value, you see:

~~~scala
"metadata":{["ml_attr":{"attrs":{"numeric":[{"idx":0,"name":"v1"},{"idx":1,"name":"v2"}]},"num_attrs":2}]}
~~~

from the output. So I think the correct TODO message is "also check metadata".

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 22 above)
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: `assert(result.collect() === expected.collect())` works for me. Note that `===` works but `==` doesn't.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 12 above)
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: Whether to test private class or not might result much longer discussion:) In MLlib, usually we expose few public APIs, while the implementation might consist of several pieces. It is useful to test each piece individually though they are not public. For example, in ALS, https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala#L53, it is hard to make useful unit test without unit testing individual components.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+      case StringType =>
+        new StringIndexer()
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: It might be necessary to implement `RModelFormula` as an `Estimator`. Otherwise, this StringIndexer() will be called every time when `transform` is called. If the input dataset is different, it would result different answers. For this PR, how about removing support for string labels. In a follow-up PR, we can make `RModelFormula` as an `Estimator`, whose `fit` returns a `RModelFormulaModel` ... (The name is awkward. Maybe we should call `RFormula` and `RFormulaModel` instead.)

**Comment by mengxr** [comment]
@ericl I make another pass. The major issue is actually that `RModelFormula` should be an `Estimator` instead of a `Transformer` in order to handle String columns. It requires some changes to the current implementation. So I would suggest removing the support for string labels in this PR and address it in a follow-up PR, since we already reviewed most of the code. It is okay to just comment out the test. Does it sound good to you?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 17 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Kept as get(), since toString should not throw.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 19 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: I added a check for this case, but kept the defaults as "feature" and "label" unless you think we should always randomize.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 21 above)
Comment (ericl) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: I see. Is the metadata important (should we include it in transformSchema)?

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 22 above)
Comment (ericl) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: == works for me, with the expected diffs?

**Comment by ericl** [comment]
@mengxr That makes sense, I'll do that in a followup PR. I also addressed the comments.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala (hunk 22 above)
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RModelFormulaSuite.scala]: `===` doesn't require `toSeq` to work. I think it is useful to use `===` everywhere in tests, just to make the code consistent. We can do this in next PR.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 16 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Actually, I mean `featureTransformer.transform` -> `transformFeatures`. This is minor.

**Comment by mengxr** [comment]
LGTM except some minor comments, which we can fix in the next PR. Merged into master. Thanks! As the next step, we can create a wrapper for `RFormula + LinearRegression` on the Scala side and then call it in R. Independently, we can add features to `RModelParser`. I'd recommend the former first in order to have some working MLlib + SparkR features in 1.5.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala (hunk 16 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala]: Ah, the problem is that featureTransformer is used for both transform and transformSchema (and I think we'll need it to transform the input data to predict).

**Comment by ericl** [comment]
Sounds good, I'll look at the R integration next.
